allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -1,18 +1,66 @@
-from allennlp.data.dataset import InstanceCollection
+from typing import Iterable, Iterator, Callable
+
 from allennlp.data.instance import Instance
 from allennlp.common import Params
 from allennlp.common.registrable import Registrable
+from allennlp.common.util import ensure_list
+
+class _LazyInstances(Iterable):
+    """
+    An ``Iterable`` that just wraps a thunk and calls it for
+    each call to ``__iter__``.
+    """
+    def __init__(self, thunk: Callable[[], Iterator[Instance]]) -> None:
+        super().__init__()
+        self.thunk = thunk
 
+    def __iter__(self) -> Iterator[Instance]:
+        return self.thunk()
 
 class DatasetReader(Registrable):
     """
-    A ``DatasetReader`` reads data from some location and constructs a :class:`Dataset`.  All
-    parameters necessary to read the data apart from the filepath should be passed to the
-    constructor of the ``DatasetReader``.
+    A ``DatasetReader`` knows how to turn a file containing a dataset into a collection
+    of ``Instance`` s.  To implement your own, just override the `_read(file_path)` method
+    to return an ``Iterable`` of the instances. This could be a list containing the instances
+    or a lazy generator that returns them one at a time.
+
+    All parameters necessary to _read the data apart from the filepath should be passed
+    to the constructor of the ``DatasetReader``.
     """
-    def read(self, file_path: str) -> InstanceCollection:
+    lazy = False
+
+    def instances(self, file_path: str) -> Iterable[Instance]:
+        """
+        Returns an ``Iterable`` containing all the instances
+        in the specified dataset.
+
+        If ``self.lazy`` is False, this calls ``self._read()``,
+        calls ``ensure_list`` on the result (which is a no-op if
+        self._read() returns a list, which most of the existing
+        dataset readers do), and returns the resulting list.
+
+        If ``self.lazy`` is True, this returns an object whose
+        ``__iter__`` method calls ``self._read()`` each iteration.
+        Note that if your implementation of ``_read()`` is not lazy
+        (i.e. it loads all instances into memory at once), then your
```

Comment:
This is a minor distinction, maybe, but what if we have `_read` _always_ return a generator, and change the `lazy` flag to a `cache` flag.  It seems semantically a little cleaner, as it would avoid this issue where `lazy` doesn't do what you think if you didn't implement the reader the right way.

Reply:
I thought about that, the things I didn't like is that (1) it's more of a breaking change, and (2) generators are "advanced" python, so I didn't want to force people to use them

Reply:
Actually, because it's a different API for `_read`, and `read` still has the same API, I don't think it breaks anything, right?

You're second point is a good one, though...  Hmm...  What if instead of having this warning in the docstring, you check for whether the result of `_read` is already a list `if self.lazy`?  If it's already a list, you could log a warning and do the same thing as in the non-lazy case.

Reply:
Also, scratch that "read still has the same API" business - I forgot for a minute that you renamed it to `instances`.

Reply:
I actually called the public method `instances` here. I am a little bit torn: if I leave it as `read` then current code will continue working. But it will sort of continue working "for the wrong reasons"

Reply:
I could go either way too on `read` versus `instances`, I wasn't suggesting here that you change it back.